fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,431 @@
+
+    committee_weight = total_active_balance // SLOTS_PER_EPOCH
+    if start_epoch == end_epoch:
+        num_committees = end_slot - start_slot + 1
+    else:
+        # A range that spans an epoch boundary, but does not span any full epoch
+        # needs pro-rata calculation
+        # First, calculate the number of committees in the current epoch
+        num_slots_in_current_epoch = (end_slot % SLOTS_PER_EPOCH) + 1
+        # Next, calculate the number of slots remaining in the current epoch
+        remaining_slots_in_current_epoch = SLOTS_PER_EPOCH - num_slots_in_current_epoch
+        # Then, calculate the number of slots in the previous epoch
+        num_slots_in_previous_epoch = SLOTS_PER_EPOCH - (start_slot % SLOTS_PER_EPOCH)
+
+        # Each committee from the previous epoch only contributes a pro-rated weight
+        multiplier = remaining_slots_in_current_epoch / SLOTS_PER_EPOCH
+        num_committees = num_slots_in_current_epoch + num_slots_in_previous_epoch * multiplier
+    return num_committees * committee_weight
+```
+
+#### `is_one_confirmed`
+
+```python
+def is_one_confirmed(store: Store, confirmation_byzantine_threshold: int, block_root: Root) -> bool:
+    current_slot = get_current_slot(store)
+    block = store.blocks[block_root]
+    parent_block = store.blocks[block.parent_root]
+    support = int(get_weight(store, block_root))
+    maximum_support = int(get_committee_weight_between_slots(store, Slot(parent_block.slot + 1), current_slot))
+    proposer_score = int(get_proposer_score(store))
+
+    return (
+        support / maximum_support >
+        0.5 * (1 + proposer_score / maximum_support) + confirmation_byzantine_threshold / 100
+    )
+```
+
+#### `is_lmd_confirmed`
+
+```python
+def is_lmd_confirmed(store: Store, confirmation_byzantine_threshold: int, block_root: Root) -> bool:
+    if block_root == store.finalized_checkpoint.root:
+        return True
+    else:
+        block = store.blocks[block_root]
+        finalized_block = store.blocks[store.finalized_checkpoint.root]
+        if block.slot <= finalized_block.slot:
+            # This block is not in the finalized chain.
+            return False
+        else:
+            # Check is_one_confirmed for this block and is_lmd_confirmed for the preceding chain.
+            return (
+                is_one_confirmed(store, confirmation_byzantine_threshold, block_root) and
+                is_lmd_confirmed(store, confirmation_byzantine_threshold, block.parent_root)
+            )
+```
+
+#### `get_remaining_weight_in_epoch`
+
+```python
+def get_remaining_weight_in_epoch(store: Store, current_slot: Slot) -> Gwei:
+    # Returns the total weight of votes for this epoch from future committees after the current slot
+    first_slot_next_epoch = compute_start_slot_at_epoch(Epoch(compute_epoch_at_slot(current_slot) + 1))
+    return get_committee_weight_between_slots(store, Slot(current_slot + 1), Slot(first_slot_next_epoch - 1))
```

Comment:
This triggers weird logic in `get_remaining_weight_in_epoch` if `current_slot == 31`.

In this case:

- `start_slot = 32`, so `start_epoch = 1`
- `end_slot = 31`, so `end_epoch = 0`

So:
 
- `start_epoch > end_epoch`

and:

- `num_slots_in_current_epoch == 32`
- `num_slots_in_previous_epoch == 32`

but with:

- `remaining_slots_in_current_epoch == 0`
- `multiplier == 0`
- `num_committees == 32`

Reply:
Great catch!

Reply:
Fixed in bb4a6aad7d0505299648a6503f9d11a3878b3175